### PR TITLE
Add CGVINodal, and share one core with CGVI

### PR DESCRIPTION
**Pull request by ZeyuanLee**
Adds a continuous Galerkin variational integrator that imposes continuity across the interval boundaries **strongly**, through an interpolatory basis, rather than weakly through Lagrange multipliers as `CGVI` does. Following [OberBloebaum:2015], the nodes of the basis are `0 = x₁ < … < x_S = 1`, so the coefficients *are* nodal values: the first is pinned to the known `qₙ`, the last *is* `qₙ₊₁`, and the momentum is computed explicitly rather than solved for. That leaves `D*(S-1)` unknowns against `CGVI`'s `D*(S+1)`, and no Lagrange multipliers.

## What changed since the original version of this PR

The first four commits are unchanged and rebased onto `main` (they were 79 commits behind). Three things on top:

**1. It is named `CGVINodal`, and the implementation is the one from NonlinearIntegrators.**

The same integrator has been maintained in [NonlinearIntegrators.jl](https://github.com/JuliaGNI/NonlinearIntegrators.jl) since this PR was opened, as the linear reference its network integrators are compared against. That version is the one taken here, because it fixes a real bug and uses the current interface:

* **`D > 1` was broken.** `components!` read `x[D*(d-1)+s]` instead of `x[D*(s-1)+d]`, so at `D = 2, S = 4` it read `x[3]` twice and `x[6]` never — a zero Jacobian column and a `SingularException` on the very first step — and `update!` broadcast one scalar across all `D` components of `q`. Both are fixed, and there is a regression test (below).
* `integrate_step!` called `solve!(x, solver, state, args)`, which `main` no longer has; it is now `solve_with_status!` + `check_solver_status`, like every other integrator.

`CGVI_II` is renamed accordingly — the name now says what distinguishes it rather than that it came second.

**2. No duplication with `CGVI`.**

Of the original 313 lines, roughly 200 were byte-identical to `integrators_cgvi.jl`: the constructor, the accessors, the traits, `show`, the cache, the initial guess, the `Q`/`V`/`P`/`F` assembly, the `residual!` dispatcher, `update!` and `integrate_step!`. Only the choice of unknowns and the three equations that follow from it are actually specific to the new integrator.

`src/integrators/cgvi/integrators_cgvi_common.jl` now holds everything the two share, behind an abstract `CGVIMethod` supertype — the same shape the DGVI family already uses in `integrators_dgvi_common.jl`:

| in the common core | per variant |
|---|---|
| traits, `default_solver`, `default_iguess` | `description` |
| `basis`, `quadrature`, `nbasis`, `nnodes` | `solversize` |
| `cgvi_coefficients`, `show` | `initial_guess!` |
| `CGVICache`, `Cache`, `CacheType`, `nlsolution` | `components!` |
| `initial_guess_at!`, `initial_guess_positions!` | `residual!` |
| `components_q!`, `components_v!`, `components_p!` | `update!` |
| `residual!` dispatcher, `update!`, `integrate_step!` | |

`CGVI` is unchanged in **results** — on the harmonic oscillator it still returns `q = 0.380122298537957` and `p = -0.22968134246613195` at the final step, bit for bit — and in **cost**: 173,936 B / 0.63 ms against `main`'s 174,240 B / 0.57 ms. Getting there took a second pass; see the last commit, which is worth reading if you review the cache.

`CGVICache` is shared by both variants. It loses `s̃`, which was allocated and never read, and takes the degrees of freedom and the solver size as *constructor arguments* rather than type parameters — `CacheDict` type-asserts its `getindex` on `CacheType(ST, problem, method)`, so that expression must constant-fold or the cache infers abstractly and every `C.X[j][k]` in the Newton hot path boxes.

Both variants now define `solversize`, as every other implicit method here does. Previously the CGVI family took the `solversize == 0` fallback, so `initsolver` gave it an `8eps` `f_abstol` floor no matter how large its system was.

**3. The interpolatory-basis requirement is enforced, not just documented.**

The formulation reads `q(0)` off `X[1]` and `q(1)` off `X[S]` instead of reconstructing them. On a Gauss basis that is simply wrong, and it fails silently — the integrator builds, runs, and returns a wrong trajectory. The constructor now rejects it and reports what it got, as `DGVIP0` already does for its own precondition.

## Tests

The separate `galerkin_integrators_II_tests.jl` asserted one number about the new integrator and nothing about the existing one. Both variants are now exercised side by side in `galerkin_integrators_tests.jl`, so a property asserted for one is asserted for the other, and `runtests.jl` needs no extra entry.

| check | before | now |
|---|---|---|
| harmonic-oscillator accuracy | `CGVI` only, `q` only | both, `q` and `p`, measured tolerances |
| `D = 2` `CoupledHarmonicOscillator` regression | neither | both — this is what catches the flat-vector layout bug, which collapses to the identity at `D = 1` |
| trait sweep | neither | both |
| `basis`/`quadrature`/`nbasis`/`nnodes` | neither | both |
| reduced precision (`Float32`, no silent promotion) | neither | both |
| bad-basis rejection | neither | `CGVINodal` |
| `show` | `CGVI` | both |
| convergence order | `CGVI(Gauss(s))` | + `CGVINodal(Lobatto(s))` |

Measured orders for `CGVINodal(Lobatto(s))` are **2.005, 4.001, 6.02** at s = 2, 3, 4 — the same `2s-2` as `CGVI(Gauss(s))`, which is why the two families assert against the same numbers.

Full suite green, docs build clean.

## Related

NonlinearIntegrators removes its copy in [JuliaGNI/NonlinearIntegrators.jl#83](https://github.com/JuliaGNI/NonlinearIntegrators.jl/pull/83). The two **merge** independently — NonlinearIntegrators gains no dependency on this package. The **releases** are ordered: `CGVINodal` was exported from NonlinearIntegrators v0.2.0, so a GeometricIntegrators release containing this PR should exist before NonlinearIntegrators tags the one that removes it, or the migration path is briefly unavailable.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/integrators/cgvi/integrators_cgvi_common.jl [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+
+function Cache{ST}(problem::AbstractProblemIODE, method::CGVIMethod; kwargs...) where {ST}
+    D = length(vec(initial_conditions(problem).q))
+    CGVICache{ST,D,nbasis(method),nnodes(method),nunknowns(method, D)}(; kwargs...)
+end
+
+@inline function CacheType(ST, problem::AbstractProblemIODE, method::CGVIMethod)
+    D = length(vec(initial_conditions(problem).q))
+    CGVICache{ST,D,nbasis(method),nnodes(method),nunknowns(method, D)}
+end
```
Comment (michakraus) [inline src/integrators/cgvi/integrators_cgvi_common.jl]: **This is where the `CGVI` regression comes from.**

`CacheDict`'s `getindex` type-asserts its return on `CacheType(ST, c.problem, c.method)` (GeometricIntegratorsBase `src/cache.jl:43-50`). On `main` that expression read nothing but type parameters and folded to a constant. Now it computes `D` from a *value* and calls `nunknowns(method, D)`, so it no longer folds:

```
CGVICache{Float64, _A, 4, 4} where _A     # here
CGVICache{Float64, 4, 4}                  # main
```

With the cache inferred abstractly, every `C.X[j][k]` in the Newton hot path allocates a box. The allocation profile is unambiguous — 264 KB and 10,752 allocations from `components_q!` line 218 alone, and the same again from `components_v!` line 234.

Keeping the cache parameterised on `{ST,S,R}` as `main` had it, and taking `D` and the solver size as constructor *arguments*, restores it:

```julia
struct CGVICache{ST,S,R} <: IODEIntegratorCache{ST}
    ...
    function CGVICache{ST,S,R}(D::Int, N::Int) where {ST,S,R}
```
```julia
ndofs(cache::CGVICache) = length(cache.q̃)

@inline CacheType(ST, ::AbstractProblemIODE, method::CGVIMethod) =
    CGVICache{ST,nbasis(method),nnodes(method)}
```

The `CGVICache` docstring above needs the same treatment — it currently documents the `N` type parameter.

File: src/integrators/cgvi/integrators_cgvi_common.jl [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+            for j in eachindex(C.X)
+                y += M.a[i, j] * C.X[j][k]
+            end
+            C.V[i][k] = y / timestep(int)
+        end
+    end
+end
+
+"Compute the one-form `P = ϑ(Q,V)` and the forces `F = f(Q,V)` at the quadrature nodes."
+function components_p!(sol, params, int::GeometricIntegrator{<:CGVIMethod}, ST)
```
Comment (michakraus) [inline src/integrators/cgvi/integrators_cgvi_common.jl]: `ST` arrives as an untyped value argument, and `components_p!` does nothing with it but forward it to `cache(int, ST)`. That is exactly the "argument of type `Type` merely passed through" case in which Julia declines to specialise, so the body is compiled once, generically.

`..., ::Type{ST}) where {ST}` on all three fixes it. On its own it takes `CGVI` from 834 KB down to 212 KB; combined with the `CacheType` change it reaches parity with `main`.

While here: `components_q!` and `components_v!` take `sol` and `params` and use neither.

**Review by michakraus — COMMENTED** [review]
`CGVINodal` itself looks right to me. The `D = 2` regression test and the three new convergence rows are exactly the two things this needed, and I additionally checked that the assertion does not over-constrain: `CGVINodal(Lagrange(nodes(Lobatto(4))), GaussLegendreQuadrature(4))` builds and integrates fine, as does a Lobatto(3) basis against Gauss(4) quadrature. The `D > 1` fix and the move off `solve!` are both clearly worth having here.

The refactor, though, is not free for `CGVI`. The results are unchanged bit for bit as claimed, but the cost is not:

| tree | allocated | time |
|---|---|---|
| `main` | 174,240 B | 0.57 ms |
| this PR | 834,144 B | 1.47 ms |

`iodeproblem()` with `CGVI(Lagrange(nodes(Gauss(4))), Gauss(4))`, minimum of 20 runs for allocations and 50 for time. That is 4.8× the allocations and 2.6× the runtime of `main`, for an integrator the PR does not intend to change. `CGVINodal` pays the same tax — 1,037,728 B.

Three things cause it, detailed inline. All three are fixed in the follow-up commit, which brings `CGVI` back to 173,936 B / 0.65 ms and `CGVINodal` down to 172,784 B.

### Summary of the inline comments

1. **The cache type is no longer inferrable.** `CacheType` now folds a value-derived `D` and `nunknowns(method, D)` into the cache's type parameters, so `cache(int, ST)` infers as `CGVICache{Float64, _A, 4, 4} where _A` where `main` got `CGVICache{Float64, 4, 4}`, and the Newton hot path boxes on every `C.X[j][k]`. This is the bulk of the regression.
2. **`components_q!`/`components_v!`/`components_p!` take `ST` as an untyped value argument.** `components_p!` only forwards it to `cache(int, ST)`, which is precisely the case where Julia declines to specialise on a `Type` argument.
3. **`nunknowns` duplicates `solversize`**, which the framework already defines for exactly this purpose and which feeds the default `f_abstol`.

Plus a handful of small ones: the `CGVINodal` docstring currently claims `CGVI` needs no Lagrange multipliers, two docstrings say "boundary conditions" where only the initial condition applies, `components_q!`/`components_v!` carry unused `sol`/`params`, the `NDOF` type parameter now collides conceptually with the newly-introduced `ndofs`, `residual!` mixes `eachindex(p̄)` and `1:D` for the same bound in adjacent blocks, the assertion does not report the offending values the way the `DGVIP0` precedent does, and two test expressions `collect` a whole time series to read its last element.

### One thing for a separate PR

`integrators_dgvi_common.jl:385` has the identical `CacheType` shape and the identical symptom — `DGVICache{Float64, _A, 4, 4, 0} where _A` on `main` too. This PR copied a pattern that was already costing the DGVI family; worth its own fix.

File: src/integrators/cgvi/integrators_cgvi_common.jl [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+nnodes(::CGVIMethod{T,S,R}) where {T,S,R} = R
+
+"""
+Number of unknowns of the nonlinear system for a problem with `D` degrees of freedom.
+
+This is what distinguishes the two variants at the level of the cache: [`CGVI`](@ref)
+solves for all `S` basis coefficients plus the endpoint momentum, [`CGVINodal`](@ref)
+for the `S-1` coefficients that the boundary conditions leave free.
+"""
+function nunknowns end
```
Comment (michakraus) [inline src/integrators/cgvi/integrators_cgvi_common.jl]: This is `solversize`, which the framework already has. It is imported into `Integrators` at `src/Integrators.jl:47`, and DGVI, VPRK, IPRK, IRK, DIRK, FLRK and PGLRK all define it for precisely this quantity.

It is not only a naming duplication. `initsolver` derives its default tolerance from it:

```julia
f_abstol = max(8, solversize(method, problem)) * eps(datatype(problem))
```

Neither CGVI variant defines `solversize`, so both take the `solversize(::GeometricMethod, ::GeometricProblem) = 0` fallback and get an `f_abstol` floor of `8eps` regardless of system size — which is the very thing the new `Float32` test has to work around by hand.

Mirroring `integrators_dgvi_common.jl:82`:

```julia
solversize(method::CGVI, problem::AbstractProblemIODE) =
    length(vec(initial_conditions(problem).q)) * (nbasis(method) + 1)

solversize(method::CGVINodal, problem::AbstractProblemIODE) =
    length(vec(initial_conditions(problem).q)) * (nbasis(method) - 1)
```

This does not disturb the bit-for-bit claim on the tested problems: at `D = 1, S = 4` both sizes are below 8, so `max(8, ·)` is unchanged.

Also, the docstring's "the `S-1` coefficients that the boundary conditions leave free" is a touch off — only the *initial* condition pins a coefficient. `X_S` is free; it is what *becomes* `q_{n+1}`. Same wording in `initial_guess_positions!` below.

File: src/integrators/cgvi/integrators_cgvi_nodal.jl [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+In the Galerkin framework of [OberBloebaum:2015](@cite) the continuity of the trajectory at
+the interval boundaries is enforced by the approximation polynomial itself rather than by a
+constraint added to the action: the basis is interpolatory with nodes
+``0 = x_1 < x_2 < \dots < x_S = 1``, so that ``\varphi_1(0) = 1`` and ``\varphi_S(1) = 1``
+and the coefficients *are* nodal values. The first is therefore pinned to the known ``q_n``,
+the last *is* ``q_{n+1}``, and the momentum is computed explicitly rather than solved for.
+
+That leaves `D*(S-1)` unknowns in the nonlinear system, against the `D*(S+1)` of
+[`CGVI`](@ref), which solves for all `S` coefficients plus the endpoint momentum and needs
+no Lagrange multipliers here.
```
Comment (michakraus) [inline src/integrators/cgvi/integrators_cgvi_nodal.jl]: This says the opposite of what is meant: `CGVI` is the variant that *does* use Lagrange multipliers. The trailing clause has attached itself to the wrong subject. Splitting the sentence fixes it:

> That leaves `D*(S-1)` unknowns in the nonlinear system, against the `D*(S+1)` of [`CGVI`](@ref), which solves for all `S` coefficients plus the endpoint momentum. No Lagrange multipliers are needed here.

File: src/integrators/cgvi/integrators_cgvi_nodal.jl [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+        # The formulation reads q(0) off X[1] and q(1) off X[S] instead of reconstructing
+        # them, which is only valid when φ₁(0) = φ_S(1) = 1 and every other basis function
+        # vanishes there. A basis on interior nodes passes every other check and produces
+        # a wrong trajectory, so it is rejected here rather than only documented.
+        e₀ = SVector{NBASIS,T}(ntuple(i -> T(i == 1), NBASIS))
+        e₁ = SVector{NBASIS,T}(ntuple(i -> T(i == NBASIS), NBASIS))
+        @assert isapprox(coeffs.r₀, e₀; atol=sqrt(eps(T))) && isapprox(coeffs.r₁, e₁; atol=sqrt(eps(T))) (
+            "CGVINodal requires an interpolatory basis with nodes at both ends of the interval, " *
+            "e.g. Lagrange(QuadratureRules.nodes(LobattoLegendreQuadrature(s))).")
```
Comment (michakraus) [inline src/integrators/cgvi/integrators_cgvi_nodal.jl]: The `DGVIP0` precedent this cites interpolates the offending values into the message (`"got $S and $R"`); this one leaves the user to work out which end failed. Worth adding `r₀` and `r₁` to the message.

Unrelated, two lines of style further down: `residual!` uses `eachindex(p̄)` at line 124 and `1:D` at line 135 for the same bound in adjacent blocks.

File: src/integrators/cgvi/integrators_cgvi_nodal.jl [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+* `b`: weights of the quadrature rule
+* `c`: nodes of the quadrature rule
+* `x`: nodes of the basis
+* `m`: mass matrix
+* `a`: derivative matrix
+* `r₀`: reconstruction coefficients at the beginning of the interval
+* `r₁`: reconstruction coefficients at the end of the interval
+
+"""
+struct CGVINodal{T,NBASIS,NNODES,NDOF,basisType<:Basis{T}} <: CGVIMethod{T,NBASIS,NNODES}
```
Comment (michakraus) [inline src/integrators/cgvi/integrators_cgvi_nodal.jl]: `NDOF` is `NBASIS*NNODES`, the `SMatrix` length — not a number of degrees of freedom. That was harmless while it was a private name in `CGVI`, but this PR introduces `ndofs`/`D` in the same files meaning the actual degrees of freedom, so the two now read as the same thing. Worth renaming in both structs while they are being touched.

File: test/integrators/galerkin_integrators_tests.jl [hunk 7]
```diff
@@ -12,11 +13,121 @@ pref = exact_solution(podeproblem())
...
+            timespan=(0.0, tend), timestep=0.1, parameters=params)
+
+        m = [params.m₁, params.m₂]
+        ω = sqrt.([params.k₁, params.k₂] ./ m)
+        qexact = q₀ .* cos.(ω .* tend) .+ p₀ ./ (m .* ω) .* sin.(ω .* tend)
+        pexact = .-m .* ω .* q₀ .* sin.(ω .* tend) .+ p₀ .* cos.(ω .* tend)
+
+        for meth in (CGVI(BGau4, QGau4), CGVINodal(BLob4, QLob4))
+            chsol = integrate(chprob, meth)
+            qend = [collect(chsol.q[:, d])[end] for d in 1:2]
```
Comment (michakraus) [inline test/integrators/galerkin_integrators_tests.jl]: `collect(chsol.q[:, d])[end]` materialises the whole time series to read its last element; `chsol.q[end][d]` is the same value in O(1). Same pattern in the `Float32` block below (`collect(f32sol.q[:, 1])[end]`).

Also, the tolerance table in the comment at the top of this testset has the `CGVINodal(Lobatto(4))` row's columns off by two characters.

**Comment by michakraus** [comment]
Pushed `1fcd184c`, which applies everything from the review above.

The two performance findings and the `solversize` one are the substance:

* `CGVICache` is parameterised on `{ST,S,R}` again, and takes the degrees of freedom and the solver size as constructor arguments, so `CacheType` is back to a function of the method alone and constant-folds.
* `components_q!` / `components_v!` / `components_p!` take `::Type{ST}`.
* `nunknowns` is gone; both variants define `solversize`, which is what every other implicit method here implements and what `initsolver` scales the default `f_abstol` by. Both sizes stay under 8 at `D = 1, S = 4`, so the tested trajectories are untouched.

| | allocated | time |
|---|---|---|
| `main` | 174,240 B | 0.57 ms |
| `488142fa` | 834,144 B | 1.47 ms |
| `1fcd184c` | 173,936 B | 0.63 ms |

`CGVINodal` is at 172,784 B, down from 1,037,728 B. Both caches infer concretely again (`CGVICache{Float64, 4, 4}`), and `solversize` agrees with the actual solution-vector length for both — 5 and 3.

The rest are as described inline: the docstring that had `CGVI` needing no Lagrange multipliers, the two "boundary conditions" wordings, the unused `sol`/`params`, `NDOF` → `NMA`, the `eachindex(p̄)`/`1:D` mismatch, the assertion message, and the two `collect(...)[end]` test expressions.

Full `Pkg.test("GeometricIntegrators")` green locally.
